velox/functions/prestosql/ArrayShuffle.cpp

```diff
@@ -0,0 +1,258 @@
+  T v1 = vector->valueAt(index1);
+  T v2 = vector->valueAt(index2);
+  if constexpr (std::is_same_v<T, StringView>) {
+    vector->setNoCopy(index1, v2);
+    vector->setNoCopy(index2, v1);
+  } else {
+    vector->set(index1, v2);
+    vector->set(index2, v1);
+  }
+}
+
+template <TypeKind kind>
+void applyScalarType(
+    const SelectivityVector& rows,
+    const ArrayVector* inputArray,
+    exec::EvalCtx& context,
+    VectorPtr& resultElements) {
+  using T = typename TypeTraits<kind>::NativeType;
+
+  // Copy array elements to new vector.
+  const VectorPtr& inputElements = inputArray->elements();
+  VELOX_DCHECK(kind == inputElements->typeKind());
+  const SelectivityVector inputElementRows =
+      toElementRows(inputElements->size(), rows, inputArray);
+  const vector_size_t elementsCount = inputElementRows.size();
+  resultElements =
+      BaseVector::create(inputElements->type(), elementsCount, context.pool());
+  resultElements->copy(
+      inputElements.get(), inputElementRows, /*toSourceRow=*/nullptr);
+  auto flatResults = resultElements->asFlatVector<T>();
+
+  rows.applyToSelected([&](vector_size_t row) {
+    const auto size = inputArray->sizeAt(row);
+    const auto offset = inputArray->offsetAt(row);
+    if (size == 0) {
+      return;
+    }
+    // create a shuffled index list to ensure values and nulls are shuffled
+    // together
+    std::mt19937 randGen(std::random_device{}());
+    std::vector<vector_size_t> indices(size);
+    std::iota(indices.begin(), indices.end(), 0);
+    std::shuffle(indices.begin(), indices.end(), randGen);
+
+    for (vector_size_t i = 0; i < size; ++i) {
+      swap(flatResults, offset + i, offset + indices[i]);
+    }
+  });
+}
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+template <TypeKind T>
+class ArrayShuffleFunction : public exec::VectorFunction {
+ public:
+  ArrayShuffleFunction() {}
+
+  bool isDeterministic() const override {
+    return false;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /*outputType*/,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 1);
+    auto& arg = args[0];
+    VectorPtr localResult;
+
+    // Input can be constant or flat.
+    if (arg->isConstantEncoding()) {
+      auto* constantArray = arg->as<ConstantVector<ComplexType>>();
+      const auto& flatArray = constantArray->valueVector();
+      const auto flatIndex = constantArray->index();
+
+      exec::LocalSelectivityVector singleRow(context, flatIndex + 1);
+      singleRow->setValid(flatIndex, true);
+      singleRow->updateBounds();
+
+      localResult = applyFlat(
+          *singleRow.get(), flatArray.get()->as<ArrayVector>(), context);
+      localResult =
+          BaseVector::wrapInConstant(rows.size(), flatIndex, localResult);
+    } else {
+      localResult = applyFlat(rows, arg->as<ArrayVector>(), context);
```

Comment:
TODO: need to support dictionary encoding to deal with current CI fuzzer failure.